File: src/ObjectManager/Object.Tes/FilePacks/Records/300-SNDG.Sound Generator.cs
using OA.Core;
using System;

namespace OA.Tes.FilePacks.Records
{
    public class SNDGRecord : Record
    {
        public enum SNDGType : uint
        {
            LeftFoot = 0,
            RightFoot = 1,
            SwimLeft = 2,
            SwimRight = 3,
            Moan = 4,
            Roar = 5,
            Scream = 6,
            Land = 7,
        }

        public override string ToString() => $"SNDG: {EDID.Value}";
        public STRVField EDID { get; set; } // Name
        public IN32Field DATA; // Sound Type Data
        public STRVField SNAM; // Sound ID
        public STRVField? CNAM; // Creature name (optional)

        public override bool CreateField(UnityBinaryReader r, string type, uint dataSize)
        {
            switch (type)
            {
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "DATA": DATA = new IN32Field(r, dataSize); return true;
                case "SNAM": SNAM = new STRVField(r, dataSize); return true;
                case "CNAM": CNAM = new STRVField(r, dataSize); return true;
                default: return false;
            }
        }

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, uint dataSize) => throw new NotImplementedException();
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/300-SNDG.Sound Generator.swift
//
//  SNDGRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class SNDGRecord: Record {
    public enum SNDGType: UInt32 {
        case leftFoot = 0, rightFoot, swimLeft, swimRight, moan, roar, scream, land = 7
    }

    public override var description: String { return "SNDG: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var DATA: IN32Field! // Sound Type Data
    public var SNAM: STRVField! // Sound ID
    public var CNAM: STRVField? = nil// Creature name (optional)

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        guard format == .TES3 else {
            return false
        }
        switch type {
        case "NAME": EDID = r.readSTRV(dataSize)
        case "DATA": DATA = r.readT(dataSize)
        case "SNAM": SNAM = r.readSTRV(dataSize)
        case "CNAM": CNAM = r.readSTRV(dataSize)
        default: return false
        }
        return true
    }
}
